lib/watcher-middleware.js

```diff
@@ -2,8 +2,10 @@
 
 const url = require('url');
 const path = require('path');
+const launchEditor = require('launch-editor');
```

Comment:
https://github.com/yyx990803/launch-editor/

Reply:
I'd generally prefer to use [open-editor](https://github.com/sindresorhus/open-editor) here, any reason not to?

* launch-editor is essentially based on open-editor
* open-editor seems to do the more correct thing (by shelling out for `$EDITOR`)
* open-editor does not require us to drop node 4 (therefore allowing existing users to get this neat feature without upgrading ember-cli)

Reply:
@rwjblue I tend to prefer `launch-editor`. I would assume that most people probably don't have `$EDITOR` setup and with `launch-editor` it will "just work" regardless.

Reply:
In addition to what @Turbo87 has said, launch-editor will fallback to the `$EDITOR` and `$VISUAL` if its guessing mechanism doesn't work.

Reply:
>  I would assume that most people probably don't have $EDITOR setup

I think that is a faulty assumption 😝 

> I would assume that most people probably don't have $EDITOR setup and with launch-editor it will "just work" regardless.

I don't think so. It is pretty common to have multiple editors running, and launch-editor just picks the first one (which sucks). For example, right now on my system I have `emacs`, `vim`, and `code` running, which should it pick? I _think_ based on [reading this code](https://github.com/yyx990803/launch-editor/blob/cf575715fd5bc01a78c71f7bcd703e5a942748e7/packages/launch-editor/guess.js) that it will just happen to use whichever had been open the longest (since `ps x` is ordered by process id)...

Reply:
I would argue that having multiple editors open at the same time is a relatively uncommon case... 🤔

Reply:
@rwjblue one way to open a specific editor at all times is to pass the `$EDITOR` or `$VISUAL` values to open-editor [here](https://github.com/ember-cli/broccoli-middleware/pull/47/files#diff-a104ba50da3745b6273d0533d2ffd375R50). That way, the ENV values get a higher priority and will launch-editor fall back to its guessing mechanism.

Reply:
> I would argue that having multiple editors open at the same time is a relatively uncommon case...

OK? So we just don't care about being completely incoherent in that case?

> one way to open a specific editor at all times is to pass the $EDITOR or $VISUAL values to open-editor here.That way, the ENV values get a higher priority and will launch-editor fall back to its guessing mechanism.

Seems good to me, but I'm unsure if that would pose an issue for folks with incorrectly configured `$EDITOR` (e.g. if as @Turbo87 mentioned, folks don't properly configure `$EDITOR`, using that as the default is not good...)

Reply:
> Seems good to me, but I'm unsure if that would pose an issue for folks with incorrectly configured $EDITOR

How about having another env variable like `EMBER_CLI_EDITOR`? CRA does something similar.

With this,

* If `EMBER_CLI_EDITOR` is present, open the editor it specifies.
* If `EMBER_CLI_EDITOR` is not present and `VISUAL` or `EDITOR` are present, launch-editor will open `VISUAL` or `EDITOR`.
* If none of these env variables are present, launch-editor's guessing mechanism will make things "just-work".
* If the guessing mechanism didn't work(no editor process was running...but IMHO, developers almost always have an editor open), we can display a message on the terminal saying `To set up the editor integration, set the env variable EMBER_CLI_EDITOR to an editor of your choice and restart the server`.

The disadvantage with this approach is that we will have to educate the users about this `EMBER_CLI_EDITOR` variable. If this variable is not set and `VISUAL` and `EDITOR` are incorrectly configured, users will not be able to use this feature. If launch-editor errors out because of this, the message should help people.

Reply:
@Turbo87 @rwjblue any updates on this?

Reply:
@Turbo87 @rwjblue I will appreciate it if we could come to some consensus here 😸